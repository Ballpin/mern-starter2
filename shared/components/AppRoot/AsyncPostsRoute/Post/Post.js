import React from 'react';
import PropTypes, { string } from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withJob } from 'react-jobs';
import Helmet from 'react-helmet';
import * as PostActions from '../../../../actions/posts';
import * as FromState from '../../../../reducers';

export function Post({ post }) {
  if (!post) {
    // Post hasn't been fetched yet. It would be better if we had a "status"
    // reducer attached to our posts which gave us a bit more insight, such
    // as whether the post is currently being fetched, or if the fetch failed.
    return null;
  }

  const { title, body } = post;

  return (
    <div>
      <Helmet title={`Posts - ${title}`} />

      <h1>
        {title}
      </h1>
      <div>
        {body}
      </div>
    </div>
  );
}

function mapStateToProps(state, { match }) {
  return {
    post: FromState.getPostById(state, match.params.id),
  };
}

const mapActionsToProps = {
  fetchPost: PostActions.fetch,
};

// We use the "compose" function from redux (but the lodash/ramda/etc equivalent
// would do the same), so that we can neatly attach multiple higher order
// functions to our component.
// They get attached to our component from a bottom up approach (i.e. the
// arguments of compose from right to left).
// Firstly the "withJob" is attached, indicating we want to do some async work.
// Then the redux "connect" is attached.
// This means the redux state and action will be passed through our "withJob".
// The job is meant to fire the fetching of a post.  If no post exists within
// the redux state it will fire the "fetchPost" redux-thunk action.  If you
// look at that action you will see it returns a Promise. It is a requirement
// to return a Promise when executing an asynchronous job so that the job
// runner knows when the job is complete.  You will also see that we first
// check to see if the post already exists within the state, if so we just
// return it which would then result in a synchronous execution of our component.
export default compose(
  connect(mapStateToProps, mapActionsToProps),
  withJob({
    work: ({ match, post, fetchPost }) => {
      if (post) {
        // We already have a post, just return true.
        return true;
      }

      // Execute the redux-thunk powered action that returns a Promise and
      // fetches the post.
      return fetchPost(match.params.id);
    },
    // Any time the post id changes we need to trigger the work.
    shouldWorkAgain: (prevProps, nextProps) =>
      prevProps.match.params.id !== nextProps.match.params.id,
  }),
)(Post);

Post.propTypes = {
  post: PropTypes.shape({
    title: string,
    body: string,
  }),
};

Post.defaultProps = {
  post: {},
};
