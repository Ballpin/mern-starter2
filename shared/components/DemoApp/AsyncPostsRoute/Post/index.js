import { asyncComponent } from 'react-async-component';

export default asyncComponent({
  resolve: () => System.import('./Post'),
  ssrMode: 'boundary',
  name: 'Post',
});
