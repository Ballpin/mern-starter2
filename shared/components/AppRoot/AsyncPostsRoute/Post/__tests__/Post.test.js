/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { shallow } from 'enzyme';

import { Post } from '../Post';

const FakePost = {
  title: 'Test',
  body: 'Test this post',
};

describe('<Post />', () => {
  test('renders header', () => {
    const wrapper = shallow(<Post post={FakePost} />);
    expect(wrapper.find('h1').length).toBe(1);
  });
  test('renders the snapshot', () => {
    const wrapper = shallow(<Post post={FakePost} />);
    expect(wrapper).toMatchSnapshot();
  });
});
