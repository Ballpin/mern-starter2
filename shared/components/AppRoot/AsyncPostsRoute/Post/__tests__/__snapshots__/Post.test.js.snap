// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post /> renders the snapshot 1`] = `
<div>
  <HelmetWrapper
    encodeSpecialCharacters={true}
    title="Posts - Test"
  />
  <h1>
    Test
  </h1>
  <div>
    Test this post
  </div>
</div>
`;
